### PR TITLE
Use v2 as default for Docker Compose version

### PR DESCRIPTION
**Pull request by jacobwoffenden**
Resolves https://github.com/devcontainers/features/issues/620

- Switches to use Docker Compose v2, [v1 is deprecated](https://docs.docker.com/compose/install/standalone/#:~:text=Install%20Compose%20standalone-,Important,-From%20July%202023)
- Installs binary from GitHub instead of using Pip ([source](https://docs.docker.com/compose/install/standalone/))
- Bump to v3

### CODE REVIEW

**Comment by jacobwoffenden** [comment]
@samruddhikhandale I wasn't sure if there was an automated process for bumping versions, updating the README, so I did it manually

**Comment by jacobwoffenden** [comment]
Regarding tests:

* [PR - Test Updated Features / test (docker-in-docker, debian:12)](https://github.com/devcontainers/features/actions/runs/5603716575/jobs/10250757682?pr=621)
  * I can see `bookworm` in https://github.com/microsoft/vscode-dev-containers/blob/main/script-library/shared/settings.env#L57

File: test/docker-in-docker/docker_build.sh [hunk 1]
```diff
@@ -9,8 +9,7 @@ source dev-container-features-test-lib
 check "docker-buildx" docker buildx version
 check "docker-build" docker build ./
 
-check "installs docker-compose v1 install" bash -c "type docker-compose"
-check "installs compose-switch" bash -c "[[ -f /usr/local/bin/compose-switch ]]"
```
Comment (samruddhikhandale) [inline test/docker-in-docker/docker_build.sh:13]: You would be able to revert these changes after you add back compose switch installation!

File: test/docker-in-docker/docker_build.sh [hunk 2]
```diff
@@ -9,8 +9,7 @@ source dev-container-features-test-lib
...
-
+check "installs docker-compose v2 install" bash -c "type docker-compose"
+check
```
Comment (samruddhikhandale) [inline test/docker-in-docker/docker_build.sh:13]: nit: remove?

File: test/docker-in-docker/docker_build_no_compose.sh [hunk 3]
```diff
@@ -9,8 +9,7 @@ source dev-container-features-test-lib
 check "docker-buildx" docker buildx version
 check "docker-build" docker build ./
 
-check "not installing compose skips docker-compose v1 install" bash -c "! type docker-compose"
-check "not installing compose skips compose-switch" bash -c "[[ ! -f /usr/local/bin/compose-switch ]]"
```
Comment (samruddhikhandale) [inline test/docker-in-docker/docker_build_no_compose.sh:13]: Same here! ^

File: src/docker-in-docker/install.sh [hunk 4]
```diff
@@ -244,73 +244,25 @@ echo "Finished installing docker / moby!"
...
+        find_version_from_git_tags docker_compose_version "https://github.com/docker/compose" "refs/tags/v"
     else
-        update-alternatives --set docker-compose /usr/local/bin/compose-switch
+        docker_compose_version="${DOCKER_DASH_COMPOSE_VERSION}"
     fi
+
+    echo "(*) Installing docker-compose ${docker_compose_version}..."
+
+    curl --location https://github.com/docker/compose/releases/download/v${docker_compose_version}/docker-compose-linux-${dockerComposeArch} \
+        --output /usr/local/bin/docker-compose
```
[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/docker-in-docker/install.sh:263]: Hmmm, now that I think about it --> ^ would fail if anyone passes `v2` (including test scenarios) 

I wonder if we should deprecate passing `v2` as option? 🤔 One can pass semvers instead (eg. 2 or 2.19 or 2.19.0). This would allow us to sync with other Feature options which does not contain the `v` prefix.

As we are bumping major versions, we could sneak this change in  (if we think that's needed, else we can condition on the `"refs/tags/v"` bit! ) 😬 

// cc @Chuxel @craiglpeters @bamurtaugh for thoughts on this change.

File: src/docker-outside-of-docker/README.md [hunk 5]
```diff
@@ -19,7 +19,7 @@ Re-use the host docker socket, adding the Docker CLI to a container. Feature inv
 |-----|-----|-----|-----|
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/README.md:19]: Can we make similar changes as per my previous comments for `docker-in-docker` Feature? Thanks!

File: test/docker-outside-of-docker/scenarios.json [hunk 6]
```diff
@@ -76,26 +77,6 @@
         },
         "containerUser": "vscode"
     },
-    "docker_dash_compose_v1": {
```
Comment (samruddhikhandale) [inline test/docker-outside-of-docker/scenarios.json:79]: > You would be able to revert these changes after you add back compose switch installation! 

Can we add a test scenario which validates specific version/semver changes? Thanks!

File: src/docker-in-docker/devcontainer-feature.json [hunk 7]
```diff
@@ -24,11 +24,10 @@
...
             "enum": [
                 "none",
-                "v1",
-                "v2"
+                "latest"
             ],
-            "default": "v1",
-            "description": "Default version of Docker Compose (v1 or v2 or none)"
+            "default": "latest",
+            "description": "Default version of Docker Compose (latest, none or specific version number e.g. 2.20.2)"
```
Comment (samruddhikhandale) [inline src/docker-in-docker/devcontainer-feature.json:30]:
```suggestion
            "description": "Select or enter a Docker Compose version."
```

File: src/docker-in-docker/devcontainer-feature.json [hunk 8]
```diff
@@ -24,11 +24,10 @@
             "type": "string",
             "enum": [
```
Comment (samruddhikhandale) [inline src/docker-in-docker/devcontainer-feature.json:25]:
```suggestion
            "proposals": [
```

File: src/docker-in-docker/devcontainer-feature.json [hunk 9]
```diff
@@ -24,11 +24,10 @@
...
                 "none",
-                "v1",
-                "v2"
+                "latest"
```
Comment (samruddhikhandale) [inline src/docker-in-docker/devcontainer-feature.json:27]:
```suggestion
                "latest",
                "2.20.1",
                "2.20",
                "2"
```

**Review by samruddhikhandale — COMMENTED** [review]
Thanks you so much for taking the time to contribute this PR, we really appreciate the support! ✨ 

I have left some thoughts/comments, thanks again!

File: src/docker-in-docker/devcontainer-feature.json (hunk 9 above)
Comment (samruddhikhandale) [reply src/docker-in-docker/devcontainer-feature.json:27]: Depends on https://github.com/devcontainers/features/pull/621#discussion_r1270090974

File: src/docker-in-docker/install.sh [hunk 10]
```diff
@@ -244,73 +244,25 @@ echo "Finished installing docker / moby!"
...
+        dockerComposeArch="x86_64"
+    elif [ "${architecture}" == "arm64" ]; then
+        dockerComposeArch="aarch64"
     fi
 
-    # Install docker-compose switch if not already installed - https://github.com/docker/compose-switch#manual-installation
-    current_v1_compose_path="$(which docker-compose)"
-    target_v1_compose_path="$(dirname "${current_v1_compose_path}")/docker-compose-v1"
-    if ! type compose-switch > /dev/null 2>&1; then
-        echo "(*) Installing compose-switch..."
```
[the remark this answers is not in the document]
Comment (jacobwoffenden) [reply src/docker-in-docker/install.sh:297]: I'm not sure I understand the need for `compose-switch`, based on it's description

>Compose Switch is a replacement to the Compose V1 docker-compose (python) executable. It translates the command line into Compose V2 docker compose then run the latter.

During the installation, it already installs `docker-compose-plugin` from APT which enables `$ docker compose`, the part I'm updating installs https://github.com/docker/compose a.k.a Docker Compose v2 into `/usr/local/bin/docker-compose`

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/docker-in-docker/install.sh:297]: > During the installation, it already installs docker-compose-plugin from APT which enables $ docker compose, the part I'm updating installs https://github.com/docker/compose a.k.a Docker Compose v2 into /usr/local/bin/docker-compose

Installing docker compose v2 is perfect and that's what is required. In this context of the comment, it provides the `docker compose` command (notice the space here)

> Can we keep the installation for docker compose-switch?

With the PR changes, tools/scripts which still use the `docker-compose` command (which was provided by v1) will fail. We don't have to install v1 to serve this scenario, however, we need to make sure that `docker-compose` command (notice the hyphen here) works. 

We can achieve this by either installing the docker-compose switch or by adding a new alias. Even v2 Docker Desktop supports the use of `docker-compose` command (they use aliasing, so aliasing or compose switch anything is fine). So it feels right to add this support in the Feature as well.

Feel free to raise any questions/concerns!

[the remark this answers is not in the document]
Comment (jacobwoffenden) [reply src/docker-in-docker/install.sh:297]: Sorry, maybe I didn't articulate my point properly in my previous comment. I still don't think compose-switch is required here.

During installation, either `docker-compose-plugin` or `moby-compose` is installed from APT depending on `if [ "${USE_MOBY}" = "true" ]`, this enables the use of `$ docker compose`

But also during installation, `if [ "${DOCKER_DASH_COMPOSE_VERSION}" != "none" ]`, then it installs https://github.com/docker/compose, which to my understanding is actually the same binary as the package from APT, and this is installed into `/usr/local/bin/docker-compose`, so they are functionally the same, the difference being installation method

Packaged: https://docs.docker.com/compose/install/linux/#install-using-the-repository

VS

Manual: https://docs.docker.com/compose/install/standalone/

Which actually raises the question, is this step even required? Instead, is the right approach just to add an alias `alias docker-compose='docker compose'`?

Note:

Running the installation script with defaults results in

```bash
root ➜ /workspace $ which docker
/usr/bin/docker
root ➜ /workspace $ docker compose version
Docker Compose version 2.20.2+azure-1
root ➜ /workspace $ which docker-compose
/usr/local/bin/docker-compose
root ➜ /workspace $ docker-compose version
Docker Compose version v2.20.2
```

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/docker-in-docker/install.sh:297]: Thanks for the explanation. 

> Even v2 Docker Desktop supports the use of docker-compose command (they use aliasing, so aliasing or compose switch anything is fine).

> Which actually raises the question, is this step even required? Instead, is the right approach just to add an alias alias docker-compose='docker compose'?

Yep, aliasing should be fine as well 👍

**Comment by eitsupi** [comment]
Is there any reason why this was not merged?

it seems that docker-outside-of-docker already defaults to compose V2.

https://github.com/devcontainers/features/blob/0cf690ee14279d9e7ae5e1308950d6067d3aba81/src/docker-outside-of-docker/devcontainer-feature.json#L23-L32
